### PR TITLE
Packages: nice errors for missing imports

### PR DESCRIPTION
**Pull request by ngroman**
When constant resolution fails and in packaged mode, search the possible resolution scopes for the prefix+name and find packages that start with that name. Suggest importing them. If no matches are found fall back to fuzzy name matching.

This is the first of other error handling changes to make improve UX of `--stripe-packages` mode. We also plan to add auto-correctors for these errors in future PRs.

Few other things this does:
* Expose `vector<NameRef> &fullName()` on `PackageInfo`
* Keep and expose a list of all mangled package names on `PackageDB` to make it easy to iterate over all packages

### Motivation
UX

### Test plan

See included automated tests.

cc @jvilk-stripe @aisamanra

### CODE REVIEW

File: core/packages/PackageDB.cc [hunk 1]
```diff
@@ -102,15 +112,19 @@ const PackageInfo &PackageDB::getPackageForFile(const core::GlobalState &gs, cor
...
     }
     return *it->second;
 }
 
 bool PackageDB::empty() const {
-    return packages.empty();
+    return packages_.empty();
+}
+
+const vector<core::NameRef> &PackageDB::packages() const {
```
Comment (jvilk-stripe) [inline core/packages/PackageDB.cc:126]: Question about the name of this function: Are these package namespaces?

File: resolver/SuggestPackage.cc [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    //
+    // Consider the package `Foo::Bar::Baz` that exports the name `Foo::Bar::Baz::Thing`.
+    // The following reference to that name from code in the package `Foo::Quuz`:
+    //   Foo::Bar::Baz::Thing
+    //        ^^^  The resolution error will be for `Bar` not resolving with a resolution scope
+    //             `Foo`
+    // In this case all we search with the prefix `Foo::Bar` and find that the `Foo::Bar::Baz`
+    // package matches.
+    void findPackagesWithPrefix(const vector<core::NameRef> &prefix, vector<PackageMatch> &matches) const {
+        ENFORCE(prefix.size() > 0);
```
Comment (jvilk-stripe) [inline resolver/SuggestPackage.cc:86]: Nit:

```suggestion
        ENFORCE(!prefix.empty());
```

File: resolver/SuggestPackage.cc [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        return ctx.state.packageDB();
+    }
+
+    vector<PackageMatch> findPossibleMissingImports(const ast::ConstantLit::ResolutionScopes &scopes,
+                                                    core::NameRef name) const {
+        vector<PackageMatch> matches;
+        vector<core::NameRef> prefix;
+        // Resolution scopes are always longest to shortest. Reserve additional space for the name we're
+        // looking for.
+        prefix.reserve(nameLen(ctx, scopes.front()) + 1);
```
Comment (jvilk-stripe) [inline resolver/SuggestPackage.cc:51]: This seems like a premature optimization. I'd remove this line and `nameLen`, and just let `prefix` resize naturally with the first run of `symbol2NameParts`. That reduces the duplicate code between `nameLen` and `symbol2NameParts`.

File: core/packages/PackageDB.cc [hunk 4]
```diff
@@ -121,16 +135,23 @@ const std::vector<std::string> &PackageDB::extraPackageFilesDirectoryPrefixes()
     return extraPackageFilesDirectoryPrefixes_;
 }
 
+bool PackageDB::isTestFile(const core::GlobalState &gs, const core::File &file) {
```
Comment (jvilk-stripe) [inline core/packages/PackageDB.cc:138]: Hopefully in the future this will change to just check if the file is within a test package or a normal package.

**Review by jvilk-stripe — APPROVED** [review]
These changes look fine to me. I had a few minor suggestions.

File: core/packages/PackageDB.cc (hunk 1 above)
Comment (ngroman) [reply core/packages/PackageDB.cc:126]: These are the mangled names, which I'm effectively using as refs to pass around for packages.

**Pull request merged**